sherpa/csrc/byte_util.cc

```diff
@@ -0,0 +1,212 @@
+
+#include "sherpa/csrc/log.h"
+
+namespace sherpa {
+
+ByteUtil::ByteUtil() {
+  // The table below is copy from
+  // https://github.com/k2-fsa/icefall/blob/master/icefall/byte_utils.py
+  // which is used to train byte level bpe, if you change the table in icefall
+  // you have to change the table below accordingly.
+  byte2token_ = std::vector<int16_t>(
+      {256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269,
+       270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283,
+       284, 285, 286, 287, 32,  33,  34,  35,  36,  37,  38,  39,  40,  41,
+       42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,
+       56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,  69,
+       70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,
+       84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,
+       98,  99,  100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
+       112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
+       126, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300,
+       301, 302, 303, 304, 305, 308, 309, 310, 311, 312, 313, 314, 315, 316,
+       317, 318, 321, 322, 323, 324, 325, 326, 327, 328, 330, 331, 332, 333,
+       334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347,
+       348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361,
+       362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375,
+       376, 377, 378, 379, 380, 381, 382, 384, 385, 386, 387, 388, 389, 390,
+       391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404,
+       405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418,
+       419, 420, 421, 422});
+  max_token_ = 422;  // the max number in above table
+  token2byte_ =
+      std::vector<int16_t>(max_token_ + 1, -1);  // the max token in byte2token_
+                                                 // is 422, so we set the length
+                                                 // of token2bytes_ 423.
+  for (size_t i = 0; i < byte2token_.size(); ++i) {
+    token2byte_[byte2token_[i]] = i;
+  }
+}
+
+std::string ByteUtil::Encode(const std::string &str) const {
+  std::ostringstream oss;
+  const uint8_t *p = reinterpret_cast<const uint8_t *>(str.data());
+  for (size_t i = 0; i < str.size(); ++i) {
+    oss << CodePointToUTF8String(byte2token_[p[i]]);
+  }
+  return oss.str();
+}
+
+std::string ByteUtil::Decode(const std::string &str) const {
+  std::vector<uint8_t> bytes;
+  UTF8StringToBytes(str, &bytes);
+  std::vector<int32_t> codes;
+  BytesToCodePoints(bytes.data(), bytes.size(), &codes);
+  std::ostringstream oss;
+  for (size_t i = 0; i < codes.size(); ++i) {
+    oss << CodePointToUTF8String(codes[i]);
+  }
+  return oss.str();
+}
+
+void ByteUtil::UTF8StringToBytes(const std::string &str,
+                                 std::vector<uint8_t> *bytes) const {
```

Comment:
there isn't a standard C++ way to do this, i.e. encoding the string as UTF-8 & decoding to unicode?

Reply:
The function name might be misleading. `UTF8StringToByteBPE` may be less confusing.

Reply:
I changed the name to `UTF8StringToTokensAndMapToBytes`.